sklearn/cluster/hierarchical.py

```diff
@@ -696,7 +717,15 @@ def fit(self, X, y=None):
                              " instance, got 'memory={!r}' instead.".format(
                                  type(memory)))
 
-        if self.n_clusters <= 0:
+        if (self.distance_threshold and self.n_clusters) or \
+                (not self.n_clusters and not self.distance_threshold):
+            raise ValueError("Either n_clusters (>0) or distance_threshold "
+                             "needs to be set, got n_clusters={} and "
+                             "distance_threshold={} instead. "
+                             "Please set n_clusters=None to continue.".format(
```

Comment:
All you need to do is what you first tried, only using local variables to pass around `n_clusters`, or store a new `n_clusters_` attribute at fit time.